### PR TITLE
Add integration tests for the python examples

### PR DESCRIPTION
**Pull request by mmeijerdfki**
This PR contains various changes in order to perform tests on the python examples.
Changelog:
- added tests to all basic example scripts
- light changes to the fb_helper script, mainly consisting of additional functions
- added a dictionary conversion script to more easily deal with fb objects (`fb_to_dict.py`)
- added experimental messages abstraction, for easier fb query objects variation creation (this is currently only used in the `test_gRPC_fb_getInstances.py` script)
- added a service_manager module to prepare and execute service calls in one function call (this is as well currently only used in `test_gRPC_fb_getInstances.py` )
- added tox and pytest as the python testing framework (python tests can be started using the `tox` command on the repos root)
- integrated automatic testing using Github actions

Additional points which could be objects of discussion upon other:
- integrations of the following into `seerep.util`:
  - message abstractions
  - service_manager
  - `fb_to_dict.py` and `pb_to_dict.py`
- folder structure of the tests into the examples directory
- `fb_helper.py` changes

Implicated possible to-do tasks which result through this merge:
- documentation on testing of the examples in the mkdocs
- more documentation and explanation on how to use the added abstractions and helper functionalities
- improvements on the message abstractions interface
- more flexible easier interface and implementation of `fb_to_dict.py`
- better performance on the github actions test as the seerep server is build beforehand on every test
- deduplication of `gRPC_fb_queryPointCloud.py` and its visualization functionality (because installing `open3d` on every test is not a feasible option, so I had to create a version without it as a temporary solution)

### CODE REVIEW

**Comment by mmeijerdfki** [comment]
> Some first comments & questions:
> 
>     * Can we move the test directory into the main directory? `examples/tests` doesn't make much sense IMO.
> 
>     * There are a bunch of commits that use the same commit message `github actions build test`. Could they be merged? Otherwise, please use more meaningful commit messages.
> 
>     * Why did you choose to write your own `pb_to_dict` conversion? The ProtoBuf Python API already provides this https://googleapis.dev/python/protobuf/latest/google/protobuf/json_format.html

I improved the code and commits based on your suggestions. I think using the functions provided by protobuf is the better way of conversion.
On your first bullet point, while it shouldn't be a problem moving the tests folder, doing that would imply that all tests should reside in this folder, wouldn't it? My thought was, because tests for ROS conversions and the HDF5 Interface are residing in their own respective folder next to the associated code, tests for the examples should reside where they are now as they are dependent on the code of the examples as well. In my opinion the current solution is more coherent with the current structure.

**Comment by jarkenau** [comment]
> ... but if they should be seen as a general test to the servers API it would make sense 

Yes, in my view, they are integration tests for the full server stack, with the added bonus of verifying that the examples work. Additionally, my concern is to keep the `/examples` clean to not overwhelm people who may want to try out the API. 

Another point is that all new Python files in the `util/` folder would be packaged and released into our PyPi package. I don't think that they would be useful in there, as they are tools for testing setup and execution, am I right?

**Comment by mmeijerdfki** [comment]
> > ... but if they should be seen as a general test to the servers API it would make sense
> 
> Yes, in my view, they are integration tests for the full server stack, with the added bonus of verifying that the examples work. Additionally, my concern is to keep the `/examples` clean to not overwhelm people who may want to try out the API.
> 
> Another point is that all new Python files in the `util/` folder would be packaged and released into our PyPi package. I don't think that they would be useful in there, as they are tools for testing setup and execution, am I right?

Okay, yes that is a good point. I will change the directory structure according to your suggestion in the coming days. 

Yes that was in part a temporary solution, but I thought some of those implementations might be useful to people using the utils for the API, for example using `fb_to_dict.py` for debugging or, although currently incomplete, `service_manager.py` to more easily call the services themselves as a complement to the `fb_helper.py` script.
I agree that it would make sense to move the message abstractions into their own seperate directory under `tests`. Maybe `tests/python/gRPC/utils` would be a good location.

**Comment by jarkenau** [comment]
Why do we need the extra dependencies in the tox GitHub environment? https://github.com/agri-gaia/seerep/blob/f738fae825c7cb7e6de6f5ebc05c0678c87a5186/tox.ini#L38-L48
Compared to:
https://github.com/agri-gaia/seerep/blob/f738fae825c7cb7e6de6f5ebc05c0678c87a5186/tox.ini#L21-L28
Shouldn't they already be included via the `requirments.txt` in the Docker base image?

**Comment by mmeijerdfki** [comment]
> Why do we need the extra dependencies in the tox GitHub environment?
> 
> https://github.com/agri-gaia/seerep/blob/f738fae825c7cb7e6de6f5ebc05c0678c87a5186/tox.ini#L38-L48
> 
> Compared to:
> 
> https://github.com/agri-gaia/seerep/blob/f738fae825c7cb7e6de6f5ebc05c0678c87a5186/tox.ini#L21-L28
> 
> Shouldn't they already be included via the `requirments.txt` in the Docker base image?

Yes, I did some tests and those dependencies are provided by the Docker image. I removed them from tox and streamlined the configs.

**Comment by jarkenau** [comment]
Some additional comments: 
> better performance on the github actions test as the seerep server is build beforehand on every test

I once experimented in using [ccache](https://ccache.dev/) to speed up the build time in https://github.com/agri-gaia/seerep/commit/6261e31d78d057bca02df2aa39e54d9f5725ea71 maybe it's interesting in this regard. 
Although  I think that ~7 min for all integrations tests is not too bad. 

>- documentation on testing of the examples in the mkdocs
>- more documentation and explanation on how to use the added abstractions and helper functionalities

Created #368 for that

> deduplication of gRPC_fb_queryPointCloud.py and its visualization functionality (because installing open3d on every test is not a feasible option, so I had to create a version without it as a temporary solution)

Totally agree on that, we should also the visualization function to `util/visualization`

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -46,3 +46,34 @@ jobs:
...
+    - name: Build seerep
+      run: |
+        source /opt/ros/noetic/setup.bash
+        catkin build --workspace $GITHUB_WORKSPACE
+        source $GITHUB_WORKSPACE/devel/setup.bash
+      shell: bash
+    - name: Install python test dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install tox-gh-actions
```
Comment (jarkenau) [inline .github/workflows/test.yml:75]: Using the [Setup Python Action](https://github.com/marketplace/actions/setup-python) would be a bit cleaner for setting up and installing the pip dependencies. However, nothing too important.

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+[testenv]
+basepython = python3.8
+description = run tests
+allowlist_externals = bash
+
+[testenv:build-seerep]
+description = build seerep
+commands = bash -c "source ../devel/setup.bash;\
+                    catkin build"
```
Comment (jarkenau) [inline tox.ini:15]: Something's not right here, the `devel space` contains the final build products, which are not available before building any code. Thus, if you run:

```bash
catkin clean -y 
tox 
```
it results in:

```bash
/usr/bin/bash: ../devel/setup.bash: No such file or directory
```

It generally works in our case due to the Dev-Container running `catkin build` when starting. Switching the order of commands should be a simple solution.

File: tox.ini [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+[testenv:{integration-examples,light}]
```
Comment (jarkenau) [inline tox.ini:17]: What's the purpose of having `light` has a second name for the environment configuration?

Comment (mmeijerdfki) [reply tox.ini]: > What's the purpose of having `light` has a second name for the environment configuration?

I thought of using it as an alias, to run just that environment. But after rethinking using the more descriptive name is probably better and avoids possible confusion. I now removed the `light` environment.

File: .github/workflows/test.yml (hunk 1 above)
Comment (mmeijerdfki) [reply .github/workflows/test.yml:75]: I did some tests and the setup python action seems to mess with the `PYTHONPATH` environment variable and the python modules provided by the docker container are not recognized anymore. Catkin fails to build seerep and errors in some python module is not found. But this is just guess based on the symptoms.

**Comment by mmeijerdfki** [comment]
I added a cleanup routine for the tests in cases the server cannot delete the projects, due to crashes, manual termination or other reasons. It will check if there are any leftover project files as well as older log files and prompt the user for deletion. Pytest will accept the switch `--ignore-leftovers` to disable this behaviour. The switch is used on the Github workflow.

File: .github/workflows/test.yml (hunk 1 above)
Comment (jarkenau) [reply .github/workflows/test.yml:75]: Ok then it's not worth the extra effort

File: tests/integration-examples/conftest.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    proj_info = fb_helper.createProjectRaw(
+        grpc_channel,
+        builder,
+        "testproject",
+        "2",
+        "EPSG::4326",
+        93,
+        52.2264,
+        8.0530,
+    )
```
Comment (jarkenau) [inline tests/integration-examples/conftest.py:81]: Why do we need to create two projects before each test? I'm quite surprised that this even work when using the same name.

Comment (jarkenau) [inline tests/integration-examples/conftest.py:81]: This fixture is never used

**Comment by jarkenau** [comment]
>I added a cleanup routine for the tests in cases the server cannot delete the projects, due to crashes, manual termination or other reasons. It will check if there are any leftover project files as well as older log files and prompt the user for deletion. Pytest will accept the switch --ignore-leftovers to disable this behaviour. The switch is used on the Github workflow.

Generally a really good idea. However, IMO things are more complicated than they need to be due to supporting `tox` and `pytest` for running the test. I believe we should only support one way to run the tests to keep it as straightforward as possible. For example, using `pytest` fixtures with the subprocess module could be used to set up the server without tox. I don't think we gain a lot of benefits from `tox`, as we don't test against different Python versions and the environment is mostly provided by the docker container (but I could be wrong on this, I just quickly skimmed the documentation). Do you agree on this?

File: tests/integration-examples/conftest.py (hunk 4 above)
Comment (mmeijerdfki) [reply tests/integration-examples/conftest.py:81]: Correct this was in preparation for a test for `gRPC_fb_createGeodeticCoordProject.py` but after investigating the proj pipeline in
https://github.com/agri-gaia/seerep/blob/46342baa989ef75cad2d2f29dc3784b5e8bd24b1/seerep_srv/seerep_core/src/core_project.cpp#L69-L91
while debugging the values of the returned polygon didn't make sense, if I remember correctly. I could look into this in the future and create a issue for that if necessary.

And for now I think I will remove the fixture, as it currently doesn't serve a purpose.

**Comment by mmeijerdfki** [comment]
> > I added a cleanup routine for the tests in cases the server cannot delete the projects, due to crashes, manual termination or other reasons. It will check if there are any leftover project files as well as older log files and prompt the user for deletion. Pytest will accept the switch --ignore-leftovers to disable this behaviour. The switch is used on the Github workflow.
> 
> Generally a really good idea. However, IMO things are more complicated than they need to be due to supporting `tox` and `pytest` for running the test. I believe we should only support one way to run the tests to keep it as straightforward as possible. For example, using `pytest` fixtures with the subprocess module could be used to set up the server without tox. I don't think we gain a lot of benefits from `tox`, as we don't test against different Python versions and the environment is mostly provided by the docker container (but I could be wrong on this, I just quickly skimmed the documentation). Do you agree on this?

Agreed, if no other types of python tests are planned for the future.
I'll test the viability of just using `pytest` for that job.

File: tests/integration-examples/conftest.py (hunk 4 above)
Comment (jarkenau) [reply tests/integration-examples/conftest.py:81]: Ok, if some results doesn't make sense please create an issue, otherwise we'll forget about it.

**Review by jarkenau — APPROVED** [review]

**Pull request merged**